### PR TITLE
Fixes unit moving during attack selection screen

### LINKED ISSUE #8728: Unit move during attack selection screen
### Game and System Information

- **Version:** 1.18.0/master
- **OS:** GNU/Linux (Debian 12)
- Installed with Flatpack

### Description of the bug

When you select a unit and click another unit who is not near to him to attack it, the first units will be teleported next to the second during the attack selection screen.
Maybe a video is more clear:
[Capture vidéo du 2024-04-10 17-10-00.webm](https://github.com/wesnoth/wesnoth/assets/90179203/7c8e0d1b-6ee8-45b1-baed-ffdc94a1d1d0)
Normally, you can't see it because the unit is centered on the screen, but if you place you in a corner, the bug will appear. 
It is maybe an old bug, but I don't found it in the issue tracker.

### Steps to reproduce the behavior

1. Be on a corner of the map,
2. Attack another unit,
3. See.

### Expected behavior

Normally, the unit will not move until the attack selection is complete. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Mahoyomu**
### Goal of the PR
- Resolves #8728.
### How does the PR work?
- By keeping the unit place not changed, and deal with the logic that a unit must stand beside the other in order to launch an attack.
### Does it resolve any reported issue?
- #8728
### If not a bug fix, why is this PR needed? What use cases does it solve?
- This is a bug fix.

## To do

Adjust possibly missed related method. If the ranged attack units(which can attack from 2 or more hexes away) come into being, more changes need to take place for backward compat.

This PR is Ready for Review.

## How to test
Compile and run, see the test video I uploaded, or see changes in the src.

### CODE REVIEW

File: src/actions/attack.hpp [hunk 1]
```diff
@@ -62,6 +62,7 @@ struct battle_context_unit_stats
 	bool swarm;              /**< Attack has swarm special. */
 	bool firststrike;        /**< Attack has firststrike special. */
 	bool disable;            /**< Attack has disable special. */
+	bool out_of_range;       /**< Attack has out_of_range special. */
```
Comment (CelticMinstrel) [inline src/actions/attack.hpp:65]: This comment seems wrong? Out of range is not a special ability, right?

File: src/mouse_events.cpp [hunk 2]
```diff
@@ -978,9 +978,9 @@ void mouse_handler::move_action(bool browse)
 					}
 
 					// block where we temporary move the unit
-					{
-						temporary_unit_mover temp_mover(pc_.get_units(), src, attack_from, itor->move_left, true);
-						choice = show_attack_dialog(attack_from, clicked_u->get_location());
+					{//fork 
+						//temporary_unit_mover temp_mover(pc_.get_units(), src, attack_from, itor->move_left, true);//this is used to move unit to `attackfrom`.
+						choice = show_attack_dialog(src, clicked_u->get_location(), false);
```
Comment (CelticMinstrel) [inline src/mouse_events.cpp:982]: This seems like it's going to give the wrong answer in some cases. Consider a scenario with the following conditions:

1. Red elvish archer is standing on mountains.
2. Blue elvish archer is standing on forest.
3. North of blue elf is forest.
4. South of blue elf is flat.
5. It's red's turn.

Now, there are two possibilities here (ignoring the diagonal directions to keep things simple):

1. Red attacks blue from the north. It has a 30% chance to be hit.
2. Red attacks blue from the south. It has a 60% chance to be hit.

The 30% or 60% chance _should_ be shown in the attack dialog, depending on which of the two adjecent hexes was hovered before the mouse was moved over the blue elf. But with this change, it looks like the attack dialog will instead show a 40% chance to be hit. Could you test and see if this is the case?

I think _both_ locations probably need to be passed to show_attack_dialog – that is, both `src` and `attack_from`.

**Comment by CelticMinstrel** [comment]
A couple of general notes…

1. I don't understand the random `//fork` comments. They should probably be deleted? Unless there's a good reason for them that I'm missing?
2. Don't comment out code, just delete it instead.

**Comment by babaissarkar** [comment]
Cleaned up the original comment a bit.

File: src/mouse_events.cpp (hunk 2 above)
Comment (stevecotton) [reply src/mouse_events.cpp:982]: Also ability calculations (leadership for example) depend on the hex that the unit is standing on, and there are some that affect the hexes around the unit that's moving - for example illuminates. That needs the correct hex to be passed around not just in the ability calculation code, but also into the time of day code.

The ability calculation code is already sufficiently complex that we're having trouble handling changes to it.

**Comment by gfgtdf** [comment]
The reason why the unit was moved to the tile is that te dialog must show the correct damage, defense etc) values that the unit has during the actual fight when the unit is on that tile, for example:
- the damage dialog must conside the correct leadership abilities on the destination tile
- the damage dialog must consider the correct defense values from the destination tile.
- the damage dialog must evaluate the more complex filters of abilities correctly as if the unit was on the Destination tile.

You must check that all of these still work correctly wen you are changing this code

**Comment by CelticMinstrel** [comment]
> * the damage dialog must evaluate the more complex filters of abilities correctly as if the unit was on the Destination tile.

However, I want to note that it's _probably_ not necessary for the unit to _actually be on that tile_ in order to do this – filters have a mechanism to evaluate it "as if" the unit had moved.

**Comment by stevecotton** [comment]
> However, I want to note that it's _probably_ not necessary for the unit to _actually be on that tile_ in order to do this – filters have a mechanism to evaluate it "as if" the unit had moved to a given tile.

But it has to take account of things such as the Illuminates ability, which means the time-of-day calculation has to be done with the unit on that hex, etc.

I feel that this method of solving a graphical problem by not moving the unit is going to uncover more and more problems that have to be worked around. I think it's going to be much easier to solve at the graphics layer, and more importantly, doing it at the graphics layer will be easy to test, whereas doing it with the current method is likely to have edge cases which are only found after the PR merges.

**Comment by gfgtdf** [comment]
> However, I want to note that it's probably not necessary for the unit to actually be on that tile in order to do this – filters have a mechanism to evaluate it "as if" the unit had moved to a given tile.

I think this feature works only limited and only in the filter for that specific unit. Let's say for example the unit has an abilty that increases damage when the defender is adjacent to two unit if the opposite side (like backstab but independent of the exact  position of the this unit), I kinda doubt that the said mechanism can handle that.


> I feel that this method of solving a graphical problem by not moving the unit is going to uncover more and more problems that ha...

I fully agree, I also think it shouldn't be too hard to do the computations _before_ the dialog is shown so tht the unit can be moved back to its original position before the dialog shows

**Comment by stevecotton** [comment]
> Let's say there's a tile that can cause the unit to take a debuff that, whenever the unit takes a step forward, causes the unit to take 1 hp damage, and happens to be the tile the attacker will be on. Now the temporary move can just destroy the way the debuff was supposed to work and make it very difficult to implement, because all that damage can happen before the unit even steps on the debuff tile.

That would probably be implemented as an `[event]name=enter_hex` or `[event]name=moveto`. The damage wouldn't occur for the temporary move, so the stats would be shown without the damage, but the unit would make the real move before actually attacking.

The UI would be likely to be a "move interrupted, press t to continue", followed by having to select the attack again once the unit is really standing on the hex that it's going to attack from. Whether that's a good UI or not depends on whether the damage was meant to be a hidden trap, or an obvious source of damage such as a fire.

File: src/display.hpp [hunk 3]
```diff
@@ -975,6 +976,49 @@ class display : public gui2::top_level_drawable
...
+		return false;
+	}
+
+	static bool has_locked_location()
+	{
+		return !locked_locations.empty();
+	}
+
+private:
+	static std::vector<const map_location*> locked_locations;
```
Comment (stevecotton) [inline src/display.hpp:1021]: Either this or add_locked_location ought to have documentation defining what a "locked location" is.

File: src/display.hpp [hunk 4]
```diff
@@ -975,6 +976,49 @@ class display : public gui2::top_level_drawable
 
 protected:
 	static display * singleton_;
+
+public:
+	static bool add_locked_location(const map_location* mlptr)
```
Comment (stevecotton) [inline src/display.hpp:981]: The return value seems to be ignored, so why must this return a value at all?

Creating a RAII class that handles the `remove_locked_location` in its destructor would be better - the design of temporary_unit_mover shows a RAII class.

**Comment by gfgtdf** [comment]
imo ading sicha new mechanic to the display code has a hige change to instrduce new issues, like what happens if the units are bigger than one tile, what happens if they are animated etc.


I think it woudl probably be far more easier if you woudl just move the attack caluclation code outside of the attack_dialogs constructor so that we can do the attack calculations before showing the attack dialogs so that we can move the unit back before the attack dialog is shown.

**Comment by gfgtdf** [comment]
well basicially you'd take this code https://github.com/wesnoth/wesnoth/blob/b6dead682bc4ecb996f99a06589ef337ba04a73f/src/gui/dialogs/unit_attack.cpp#L79 and move it there https://github.com/wesnoth/wesnoth/blob/b6dead682bc4ecb996f99a06589ef337ba04a73f/src/mouse_events.cpp#L1356 but have to modify it so that it stores the result in a structure instead of setting the dialog values immidiately, and then you move the unit mover from here https://github.com/wesnoth/wesnoth/blob/b6dead682bc4ecb996f99a06589ef337ba04a73f/src/mouse_events.cpp#L982 to an extra scope in here https://github.com/wesnoth/wesnoth/blob/b6dead682bc4ecb996f99a06589ef337ba04a73f/src/mouse_events.cpp#L1344

**Comment by Pentarctagon** [comment]
@stevecotton @CelticMinstrel @gfgtdf any other issues that need to be resolved?

**Comment by Pentarctagon** [comment]
@stevecotton @CelticMinstrel @gfgtdf yes/no?

**Comment by stevecotton** [comment]
Although @gfgtdf said to move stuff into mouse_events.cpp, my feeling is that we should keep dialog-specific code in the dialog-specific file, instead of adding more to the already bloated mouse_events.cpp file. However, I haven't looked at how I'd implement it, which is why it's taken me so long to comment.

There's a bug introduced here - the dialog can show a subdialog, when the "Damage Calculation" button is pressed. Put an ally with leadership next to the target, and compare the dialogs when the dialog is shown from a remote hex - the damage calculation still shows the right numbers (including leadership), but misses the "Leadership bonus +50%" line from the explanation.

**Comment by stevecotton** [comment]
Some of the calculations for that subdialog happen in src/gui/dialogs/attack_predictions.cpp, and are done when the "Damage Calculations" button is pressed, with the assumption that the unit is on the hex that it will attack from.

**Comment by gfgtdf** [comment]
we could also, instead of trying to moving the calculations of stats outside of the dialog, move the unit_mover inside these two dialogs constructors. It feels a bit strange to move a unit while the dialog is showing, but maybe thats bo so bad.

File: src/gui/dialogs/unit_attack.cpp [hunk 5]
```diff
@@ -38,13 +38,15 @@ REGISTER_DIALOG(unit_attack)
 unit_attack::unit_attack(const unit_map::iterator& attacker_itor,
 						   const unit_map::iterator& defender_itor,
 						   std::vector<battle_context>&& weapons,
-						   const int best_weapon)
+						   const int best_weapon,
+						   std::vector<gui2::widget_data>& data_vector)
```
Comment (soliton-) [inline src/gui/dialogs/unit_attack.cpp:42]: This is a really generic name. It should be changed to a name that more specifically describes what this variable holds.

File: src/gui/dialogs/unit_attack.cpp [hunk 6]
```diff
@@ -72,162 +76,8 @@ void unit_attack::pre_show()
 	keyboard_capture(&weapon_list);
 
 	// Possible TODO: If a "blank weapon" is generally useful, add it as a static member in attack_type.
```
Comment (soliton-) [inline src/gui/dialogs/unit_attack.cpp:78]: Does it make sense that this comment is left here?

Comment (soliton-) [reply src/gui/dialogs/unit_attack.cpp:78]: You moved code the comment is about, no? So either move it as well or delete it.

**Comment by Pentarctagon** [comment]
Is there anything left to do before merging this other than fixing the conflict?

File: src/gui/dialogs/unit_attack.cpp (hunk 5 above)
Comment (Mahoyomu) [reply src/gui/dialogs/unit_attack.cpp]: > This is a really generic name. It should be changed to a name that more specifically describes what this variable holds.

It's fixed in the following commit, iirc. Or the name after the following commit still couldn't describe it? If so, please share your idea with me.

**Comment by Pentarctagon** [comment]
Looks like there's a conflict that needs to be resolved, but otherwise I believe this is good to merge?

**Comment by Wedge009** [comment]
`unit_attack.cpp` - looks like a bunch of changes were done in master, not sure where that `for` loop is supposed to go...

**Comment by Mahoyomu** [comment]
I don't know if the work in this PR will still be considered acceptable in the future. I did this work a long time ago, and it's still waiting for further review...

If the work here gets approval and I'm informed that it solves the issue, or if I receive direct comments about this information, I'll do some more work to make it ready for merging. Otherwise, I think it might be better to just wait.

**Comment by Mahoyomu** [comment]
It's squashed, and the commit message is updated. I hope it will be working and mergeable this time.
